libraries/ms-helper/src/mcp2515.c

```diff
@@ -0,0 +1,235 @@
+#include "mcp2515.h"
+#include <stddef.h>
+#include <string.h>
+#include "critical_section.h"
+#include "gpio_it.h"
+#include "log.h"
+#include "mcp2515_defs.h"
+
+#define MCP2515_MAX_WRITE_BUFFER_LEN 10
+
+typedef struct Mcp2515TxBuffer {
+  uint8_t id;
+  uint8_t data;
+  uint8_t rts;
+} Mcp2515TxBuffer;
+
+typedef struct Mcp2515RxBuffer {
+  uint8_t id;
+  uint8_t data;
+  uint8_t int_flag;
+} Mcp2515RxBuffer;
+
+static const Mcp2515TxBuffer s_tx_buffers[] = {
+  { .id = MCP2515_LOAD_TXB0SIDH, .data = MCP2515_LOAD_TXB0D0, .rts = MCP2515_RTS_TXB0 },
+  { .id = MCP2515_LOAD_TXB1SIDH, .data = MCP2515_LOAD_TXB1D0, .rts = MCP2515_RTS_TXB1 },
+  { .id = MCP2515_LOAD_TXB2SIDH, .data = MCP2515_LOAD_TXB2D0, .rts = MCP2515_RTS_TXB2 },
+};
+
+static const Mcp2515RxBuffer s_rx_buffers[] = {
+  { .id = MCP2515_READ_RXB0SIDH, .data = MCP2515_READ_RXB0D0, .int_flag = MCP2515_CANINT_RX0IE },
+  { .id = MCP2515_READ_RXB1SIDH, .data = MCP2515_READ_RXB1D0, .int_flag = MCP2515_CANINT_RX1IE },
+};
+
+static void prv_reset(Mcp2515Storage *storage) {
+  uint8_t payload[] = { MCP2515_CMD_RESET };
+  spi_exchange(storage->spi_port, payload, sizeof(payload), NULL, 0);
+}
+
+static void prv_read(Mcp2515Storage *storage, uint8_t addr, uint8_t *read_data, size_t read_len) {
+  uint8_t payload[] = { MCP2515_CMD_READ, addr };
+  spi_exchange(storage->spi_port, payload, sizeof(payload), read_data, read_len);
+}
+
+static void prv_write(Mcp2515Storage *storage, uint8_t addr, uint8_t *write_data,
+                      size_t write_len) {
+  uint8_t payload[MCP2515_MAX_WRITE_BUFFER_LEN];
+  payload[0] = MCP2515_CMD_WRITE;
+  payload[1] = addr;
+  memcpy(&payload[2], write_data, write_len);
+  spi_exchange(storage->spi_port, payload, sizeof(payload), write_data, write_len);
+}
+
+static void prv_bit_modify(Mcp2515Storage *storage, uint8_t addr, uint8_t mask, uint8_t data) {
```

Comment:
Comment? This is less obvious than the other private functions as to its purpose.

Reply:
It's just calling a built-in command, but I'll add a comment.